File: FaceGuardOffline/ios/FaceGuard/TFLiteModule.swift
import Foundation
import React
import TensorFlowLite
import UIKit

@objc(TFLiteModule)
class TFLiteModule: NSObject {
  private let queue = DispatchQueue(label: "tflite", qos: .userInitiated)
  private var interpreters: [String: Interpreter] = [:]

  @objc static func requiresMainQueueSetup() -> Bool {
    false
  }

  @objc(loadModel:resolver:rejecter:)
  func loadModel(
    _ modelName: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      do {
        guard let path = Bundle.main.path(forResource: modelName.replacingOccurrences(of: ".tflite", with: ""), ofType: "tflite") else {
          throw TFLiteBridgeError.modelNotFound(modelName)
        }

        let interpreter = try self.makeInterpreter(modelPath: path)
        self.interpreters[self.modelKey(modelName)] = interpreter
        resolve("Loaded \(modelName)")
      } catch {
        reject("MODEL_LOAD_FAILED", "FaceGuard could not load \(modelName). Please reinstall the offline model pack.", error)
      }
    }
  }

  @objc(runBlazeFace:resolver:rejecter:)
  func runBlazeFace(
    _ base64Frame: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      do {
        var interpreter = try self.requireInterpreter("blazeface")
        let startedAt = CFAbsoluteTimeGetCurrent()
        let image = try self.decodeBase64Image(base64Frame)
        let input = try self.imageToFloatData(image, size: 128, mode: .minusOneToOne)
        try interpreter.copy(input, toInputAt: 0)
        try interpreter.invoke()
        let boxes = try interpreter.output(at: 0).data.toFloatArray()
        let scores = try interpreter.output(at: 1).data.toFloatArray()
        var result = self.postProcessBlazeFace(boxes: boxes, scores: scores, image: image)
        result["inferenceMs"] = self.elapsedMs(startedAt)
        resolve(try self.jsonString(result))
      } catch {
        reject("INFERENCE_FAILED", "Face detection failed. Please hold the camera steady.", error)
      }
    }
  }

  @objc(runMobileFaceNet:resolver:rejecter:)
  func runMobileFaceNet(
    _ base64Face: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      do {
        var interpreter = try self.requireInterpreter("mobilefacenet")
        let startedAt = CFAbsoluteTimeGetCurrent()
        let image = try self.decodeBase64Image(base64Face)
        let input = try self.imageToFloatData(image, size: 112, mode: .imagenet)
        try interpreter.copy(input, toInputAt: 0)
        try interpreter.invoke()
        let embedding = self.l2Normalize(try interpreter.output(at: 0).data.toFloatArray())
        resolve(try self.jsonString([
          "embedding": embedding,
          "inferenceMs": self.elapsedMs(startedAt),
        ]))
      } catch {
        reject("INFERENCE_FAILED", "Face embedding failed. Please retry with a clearer face crop.", error)
      }
    }
  }

  @objc(runMiniFASNet:resolver:rejecter:)
  func runMiniFASNet(
    _ base64Face: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      do {
        var interpreter = try self.requireInterpreter("minifasnet")
        let startedAt = CFAbsoluteTimeGetCurrent()
        let image = try self.decodeBase64Image(base64Face)
        let input = try self.imageToFloatData(image, size: 80, mode: .minusOneToOne)
        try interpreter.copy(input, toInputAt: 0)
        try interpreter.invoke()
        let output = try interpreter.output(at: 0).data.toFloatArray()
        let spoofScore = self.softmax(target: output[0], other: output[1])
        let realScore = self.softmax(target: output[1], other: output[0])
        resolve(try self.jsonString([
          "isReal": realScore >= 0.5,
          "realScore": realScore,
          "spoofScore": spoofScore,
          "spoofType": realScore >= 0.5 ? NSNull() : "presentation_attack",
          "inferenceMs": self.elapsedMs(startedAt),
        ]))
      } catch {
        reject("INFERENCE_FAILED", "Passive liveness check failed. Please retry.", error)
      }
    }
  }

  private func requireInterpreter(_ key: String) throws -> Interpreter {
    if let interpreter = interpreters[key] {
      return interpreter
    }
    if let match = interpreters.first(where: { $0.key.localizedCaseInsensitiveContains(key) })?.value {
      return match
    }
    throw TFLiteBridgeError.modelNotLoaded(key)
  }

  private func makeInterpreter(modelPath: String) throws -> Interpreter {
    var options = Interpreter.Options()
    options.threadCount = 4

    let delegateAttempts: [[Delegate]] = [
      [MetalDelegate()],
      [CoreMLDelegate()],
      [],
    ]
    var lastError: Error?

    for delegates in delegateAttempts {
      do {
        // WOW factor: MetalDelegate is attempted first so FaceGuard gets GPU-speed offline inference on iOS.
        var interpreter = try Interpreter(modelPath: modelPath, options: options, delegates: delegates)
        try interpreter.allocateTensors()
        return interpreter
      } catch {
        lastError = error
      }
    }

    throw lastError ?? TFLiteBridgeError.modelNotLoaded(modelPath)
  }

  private func modelKey(_ modelName: String) -> String {
    modelName.replacingOccurrences(of: ".tflite", with: "").lowercased()
  }

  private func decodeBase64Image(_ base64: String) throws -> UIImage {
    let payload = base64.components(separatedBy: ",").last ?? base64
    guard let data = Data(base64Encoded: payload), let image = UIImage(data: data) else {
      throw TFLiteBridgeError.invalidImage
    }
    return image
  }

  private func imageToFloatData(_ image: UIImage, size: Int, mode: NormalizeMode) throws -> Data {
    guard let cgImage = image.cgImage else {
      throw TFLiteBridgeError.invalidImage
    }

    let width = size
    let height = size
    let bytesPerPixel = 4
    var pixelBuffer: CVPixelBuffer?
    let attrs = [
      kCVPixelBufferCGImageCompatibilityKey: true,
      kCVPixelBufferCGBitmapContextCompatibilityKey: true,
    ] as CFDictionary
    let status = CVPixelBufferCreate(
      kCFAllocatorDefault,
      width,
      height,
      kCVPixelFormatType_32RGBA,
      attrs,
      &pixelBuffer
    )
    guard status == kCVReturnSuccess, let unwrappedPixelBuffer = pixelBuffer else {
      throw TFLiteBridgeError.invalidImage
    }
    let pixelBuffer = unwrappedPixelBuffer

    CVPixelBufferLockBaseAddress(pixelBuffer, [])
    defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, []) }

    guard let baseAddress = CVPixelBufferGetBaseAddress(pixelBuffer) else {
      throw TFLiteBridgeError.invalidImage
    }

    let colorSpace = CGColorSpaceCreateDeviceRGB()
    guard let context = CGContext(
      data: baseAddress,
      width: width,
      height: height,
      bitsPerComponent: 8,
      bytesPerRow: CVPixelBufferGetBytesPerRow(pixelBuffer),
      space: colorSpace,
      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    ) else {
      throw TFLiteBridgeError.invalidImage
    }

    context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))

    var floats = [Float32]()
    floats.reserveCapacity(width * height * 3)
    let bytesPerRow = CVPixelBufferGetBytesPerRow(pixelBuffer)
    let raw = baseAddress.assumingMemoryBound(to: UInt8.self)
    for y in 0..<height {
      for x in 0..<width {
        let index = y * bytesPerRow + x * bytesPerPixel
        floats.append(normalize(Float32(raw[index]), channel: 0, mode: mode))
        floats.append(normalize(Float32(raw[index + 1]), channel: 1, mode: mode))
        floats.append(normalize(Float32(raw[index + 2]), channel: 2, mode: mode))
      }
    }

    return floats.withUnsafeBufferPointer { buffer -> Data in
      guard let baseAddress = buffer.baseAddress else {
        return Data()
      }
      return Data(bytes: baseAddress, count: buffer.count * MemoryLayout<Float32>.stride)
    }
  }

  private func normalize(_ value: Float32, channel: Int, mode: NormalizeMode) -> Float32 {
    switch mode {
    case .minusOneToOne:
      return value / 127.5 - 1
    case .imagenet:
      let mean: [Float32] = [0.485, 0.456, 0.406]
      let std: [Float32] = [0.229, 0.224, 0.225]
      return (value / 255 - mean[channel]) / std[channel]
    }
  }

  private func postProcessBlazeFace(boxes: [Float32], scores: [Float32], image: UIImage) -> [String: Any] {
    var bestIndex = -1
    var bestScore: Float32 = 0
    for index in 0..<min(896, scores.count) {
      let score = sigmoid(scores[index])
      if score > bestScore {
        bestScore = score
        bestIndex = index
      }
    }

    if bestIndex < 0 || bestScore < 0.65 {
      return [
        "detected": false,
        "bbox": NSNull(),
        "keypoints": [],
        "confidence": bestScore,
        "frameWidth": image.size.width,
        "frameHeight": image.size.height,
      ]
    }

    let offset = bestIndex * 16
    let cx = boxes[offset] / 128
    let cy = boxes[offset + 1] / 128
    let width = max(Float32(0.01), boxes[offset + 2] / 128)
    let height = max(Float32(0.01), boxes[offset + 3] / 128)
    var keypoints: [[String: Float32]] = []
    for point in 0..<6 {
      keypoints.append([
        "x": clamp01(boxes[offset + 4 + point * 2] / 128),
        "y": clamp01(boxes[offset + 5 + point * 2] / 128),
      ])
    }

    return [
      "detected": true,
      "bbox": [
        "x": clamp01(cx - width / 2),
        "y": clamp01(cy - height / 2),
        "width": clamp01(width),
        "height": clamp01(height),
      ],
      "keypoints": keypoints,
      "confidence": bestScore,
      "frameWidth": image.size.width,
      "frameHeight": image.size.height,
    ]
  }

  private func l2Normalize(_ vector: [Float32]) -> [Float32] {
    let norm = sqrt(vector.reduce(Float32(0)) { $0 + $1 * $1 })
    guard norm > 1e-8 else { return vector }
    return vector.map { $0 / norm }
  }

  private func softmax(target: Float32, other: Float32) -> Float32 {
    let maxLogit = max(target, other)
    let expTarget = Foundation.exp(target - maxLogit)
    let expOther = Foundation.exp(other - maxLogit)
    return expTarget / (expTarget + expOther)
  }

  private func sigmoid(_ value: Float32) -> Float32 {
    1 / (1 + Foundation.exp(-value))
  }

  private func clamp01(_ value: Float32) -> Float32 {
    min(1, max(0, value))
  }

  private func elapsedMs(_ startedAt: CFAbsoluteTime) -> Double {
    (CFAbsoluteTimeGetCurrent() - startedAt) * 1000
  }

  private func jsonString(_ object: [String: Any]) throws -> String {
    let data = try JSONSerialization.data(withJSONObject: object, options: [])
    return String(data: data, encoding: .utf8) ?? "{}"
  }

  private enum NormalizeMode {
    case minusOneToOne
    case imagenet
  }
}

private enum TFLiteBridgeError: Error {
  case modelNotFound(String)
  case modelNotLoaded(String)
  case invalidImage
}

private extension Data {
  func toFloatArray() -> [Float32] {
    withUnsafeBytes { rawBuffer in
      Array(rawBuffer.bindMemory(to: Float32.self))
    }
  }
}

File: FaceGuardOffline/android/app/src/main/java/com/faceguard/tflite/TFLiteModule.kt
package com.faceguard.tflite

import android.content.res.AssetFileDescriptor
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import org.json.JSONArray
import org.json.JSONObject
import org.tensorflow.lite.Interpreter
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

class TFLiteModule(private val reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  private val executor: ExecutorService = Executors.newSingleThreadExecutor()
  private val interpreters = mutableMapOf<String, Interpreter>()

  override fun getName(): String = "TFLiteModule"

  @ReactMethod
  fun loadModel(modelName: String, promise: Promise) {
    executor.execute {
      try {
        val options = Interpreter.Options().apply {
          setNumThreads(4)
          try {
            setUseNNAPI(true)
          } catch (_: Throwable) {
            setUseNNAPI(false)
          }
        }
        val buffer = loadMappedModel(modelName)
        interpreters[modelKey(modelName)]?.close()
        interpreters[modelKey(modelName)] = Interpreter(buffer, options)
        promise.resolve("Loaded $modelName")
      } catch (error: Throwable) {
        promise.reject(
          "MODEL_LOAD_FAILED",
          "FaceGuard could not load $modelName. Please reinstall the offline model pack.",
          error,
        )
      }
    }
  }

  @ReactMethod
  fun runBlazeFace(base64Frame: String, promise: Promise) {
    executor.execute {
      try {
        val interpreter = requireInterpreter("blazeface")
        val startedAt = System.nanoTime()
        val bitmap = decodeBase64Jpeg(base64Frame)
        val input = bitmapToFloatBuffer(bitmap, 128, NormalizeMode.MINUS_ONE_TO_ONE)
        val boxes = Array(1) { Array(896) { FloatArray(16) } }
        val scores = Array(1) { Array(896) { FloatArray(1) } }
        interpreter.runForMultipleInputsOutputs(arrayOf(input), mapOf(0 to boxes, 1 to scores))
        val result = postProcessBlazeFace(boxes[0], scores[0], bitmap.width, bitmap.height)
        result.put("inferenceMs", elapsedMs(startedAt))
        promise.resolve(result.toString())
      } catch (error: Throwable) {
        promise.reject("INFERENCE_FAILED", "Face detection failed. Please hold the camera steady.", error)
      }
    }
  }

  @ReactMethod
  fun runMobileFaceNet(base64Face: String, promise: Promise) {
    executor.execute {
      try {
        val interpreter = requireInterpreter("mobilefacenet")
        val startedAt = System.nanoTime()
        val bitmap = decodeBase64Jpeg(base64Face)
        val input = bitmapToFloatBuffer(bitmap, 112, NormalizeMode.IMAGENET)
        val output = Array(1) { FloatArray(128) }
        interpreter.run(input, output)
        val normalized = l2Normalize(output[0])
        val json = JSONObject()
          .put("embedding", JSONArray(normalized.toList()))
          .put("inferenceMs", elapsedMs(startedAt))
        promise.resolve(json.toString())
      } catch (error: Throwable) {
        promise.reject("INFERENCE_FAILED", "Face embedding failed. Please retry with a clearer face crop.", error)
      }
    }
  }

  @ReactMethod
  fun runMiniFASNet(base64Face: String, promise: Promise) {
    executor.execute {
      try {
        val interpreter = requireInterpreter("minifasnet")
        val startedAt = System.nanoTime()
        val bitmap = decodeBase64Jpeg(base64Face)
        val input = bitmapToFloatBuffer(bitmap, 80, NormalizeMode.MINUS_ONE_TO_ONE)
        val output = Array(1) { FloatArray(2) }
        interpreter.run(input, output)
        val spoofScore = softmax(output[0][0], output[0][1])
        val realScore = softmax(output[0][1], output[0][0])
        val json = JSONObject()
          .put("isReal", realScore >= 0.5f)
          .put("realScore", realScore)
          .put("spoofScore", spoofScore)
          .put("spoofType", if (realScore >= 0.5f) JSONObject.NULL else "presentation_attack")
          .put("inferenceMs", elapsedMs(startedAt))
        promise.resolve(json.toString())
      } catch (error: Throwable) {
        promise.reject("INFERENCE_FAILED", "Passive liveness check failed. Please retry.", error)
      }
    }
  }

  override fun invalidate() {
    executor.execute {
      interpreters.values.forEach { it.close() }
      interpreters.clear()
    }
    executor.shutdown()
    super.invalidate()
  }

  private fun loadMappedModel(modelName: String): MappedByteBuffer {
    val fd: AssetFileDescriptor = reactContext.assets.openFd(modelName)
    FileInputStream(fd.fileDescriptor).use { input ->
      val channel = input.channel
      return channel.map(FileChannel.MapMode.READ_ONLY, fd.startOffset, fd.declaredLength)
    }
  }

  private fun requireInterpreter(key: String): Interpreter {
    interpreters[key]?.let { return it }
    val match = interpreters.entries.firstOrNull { it.key.contains(key, ignoreCase = true) }
    return match?.value ?: throw IllegalStateException("$key model is not loaded")
  }

  private fun modelKey(modelName: String): String =
    modelName.removeSuffix(".tflite").lowercase()

  private fun decodeBase64Jpeg(base64Frame: String): Bitmap {
    val comma = base64Frame.indexOf(',')
    val payload = if (comma >= 0) base64Frame.substring(comma + 1) else base64Frame
    val bytes = Base64.decode(payload, Base64.DEFAULT)
    return BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
      ?: throw IllegalArgumentException("Invalid JPEG frame")
  }

  private fun bitmapToFloatBuffer(bitmap: Bitmap, inputSize: Int, mode: NormalizeMode): ByteBuffer {
    val scaled = Bitmap.createScaledBitmap(bitmap, inputSize, inputSize, true)
    val buffer = ByteBuffer.allocateDirect(1 * inputSize * inputSize * 3 * 4)
    buffer.order(ByteOrder.nativeOrder())
    val pixels = IntArray(inputSize * inputSize)
    scaled.getPixels(pixels, 0, inputSize, 0, 0, inputSize, inputSize)

    for (pixel in pixels) {
      val r = (pixel shr 16) and 0xFF
      val g = (pixel shr 8) and 0xFF
      val b = pixel and 0xFF
      putNormalized(buffer, r, 0, mode)
      putNormalized(buffer, g, 1, mode)
      putNormalized(buffer, b, 2, mode)
    }
    buffer.rewind()
    return buffer
  }

  private fun putNormalized(buffer: ByteBuffer, value: Int, channel: Int, mode: NormalizeMode) {
    val normalized = when (mode) {
      NormalizeMode.MINUS_ONE_TO_ONE -> value / 127.5f - 1f
      NormalizeMode.IMAGENET -> {
        val means = floatArrayOf(0.485f, 0.456f, 0.406f)
        val stds = floatArrayOf(0.229f, 0.224f, 0.225f)
        (value / 255f - means[channel]) / stds[channel]
      }
    }
    buffer.putFloat(normalized)
  }

  private fun postProcessBlazeFace(
    boxes: Array<FloatArray>,
    scores: Array<FloatArray>,
    frameWidth: Int,
    frameHeight: Int,
  ): JSONObject {
    var bestIndex = -1
    var bestScore = 0f
    for (i in scores.indices) {
      val score = sigmoid(scores[i][0])
      if (score > bestScore) {
        bestScore = score
        bestIndex = i
      }
    }

    if (bestIndex < 0 || bestScore < 0.65f) {
      return JSONObject()
        .put("detected", false)
        .put("bbox", JSONObject.NULL)
        .put("keypoints", JSONArray())
        .put("confidence", bestScore)
        .put("frameWidth", frameWidth)
        .put("frameHeight", frameHeight)
    }

    val raw = boxes[bestIndex]
    val cx = raw[0] / 128f
    val cy = raw[1] / 128f
    val w = max(0.01f, raw[2] / 128f)
    val h = max(0.01f, raw[3] / 128f)
    val bbox = JSONObject()
      .put("x", clamp01(cx - w / 2f))
      .put("y", clamp01(cy - h / 2f))
      .put("width", clamp01(w))
      .put("height", clamp01(h))

    val keypoints = JSONArray()
    for (i in 0 until 6) {
      keypoints.put(
        JSONObject()
          .put("x", clamp01(raw[4 + i * 2] / 128f))
          .put("y", clamp01(raw[5 + i * 2] / 128f)),
      )
    }

    return JSONObject()
      .put("detected", true)
      .put("bbox", bbox)
      .put("keypoints", keypoints)
      .put("confidence", bestScore)
      .put("frameWidth", frameWidth)
      .put("frameHeight", frameHeight)
  }

  private fun l2Normalize(vector: FloatArray): FloatArray {
    var sum = 0f
    for (value in vector) sum += value * value
    val norm = sqrt(sum)
    if (norm < 1e-8f) return vector
    return FloatArray(vector.size) { vector[it] / norm }
  }

  private fun softmax(target: Float, other: Float): Float {
    val maxLogit = max(target, other)
    val expTarget = exp((target - maxLogit).toDouble()).toFloat()
    val expOther = exp((other - maxLogit).toDouble()).toFloat()
    return expTarget / (expTarget + expOther)
  }

  private fun sigmoid(value: Float): Float =
    (1f / (1f + exp((-value).toDouble()))).toFloat()

  private fun clamp01(value: Float): Float = min(1f, max(0f, value))

  private fun elapsedMs(startedAt: Long): Double =
    (System.nanoTime() - startedAt).toDouble() / 1_000_000.0

  private enum class NormalizeMode {
    MINUS_ONE_TO_ONE,
    IMAGENET,
  }
}
